crates/wasi-common/src/sys/windows/host_impl.rs

```diff
@@ -40,45 +40,70 @@ pub(crate) fn errno_from_win(error: winx::winerror::WinError) -> wasi::__wasi_er
     }
 }
 
-pub(crate) fn fdflags_from_win(mode: AccessMode) -> wasi::__wasi_fdflags_t {
+pub(crate) fn fdflags_from_win(
+    access: AccessMode,
+    mode: FileModeInformation,
+) -> wasi::__wasi_fdflags_t {
     let mut fdflags = 0;
-    // TODO verify this!
-    if mode.contains(AccessMode::FILE_APPEND_DATA) {
+
+    if access.contains(AccessMode::FILE_APPEND_DATA)
+        && !access.contains(AccessMode::FILE_WRITE_DATA)
+    {
         fdflags |= wasi::__WASI_FDFLAG_APPEND;
     }
-    if mode.contains(AccessMode::SYNCHRONIZE) {
-        fdflags |= wasi::__WASI_FDFLAG_DSYNC;
-        fdflags |= wasi::__WASI_FDFLAG_RSYNC;
+
+    if mode.contains(FileModeInformation::FILE_WRITE_THROUGH) {
+        // Only report __WASI_FDFLAG_SYNC, which is technically the only one of the O_?SYNC flags Windows supports.
         fdflags |= wasi::__WASI_FDFLAG_SYNC;
     }
-    // The NONBLOCK equivalent is FILE_FLAG_OVERLAPPED
-    // but it seems winapi doesn't provide a mechanism
-    // for checking whether the handle supports async IO.
-    // On the contrary, I've found some dicsussion online
-    // which suggests that on Windows all handles should
-    // generally be assumed to be opened with async support
-    // and then the program should fallback should that **not**
-    // be the case at the time of the operation.
-    // TODO: this requires further investigation
+
+    if !mode.contains(FileModeInformation::FILE_SYNCHRONOUS_IO_ALERT)
+        && !mode.contains(FileModeInformation::FILE_SYNCHRONOUS_IO_NONALERT)
+    {
+        fdflags |= wasi::__WASI_FDFLAG_NONBLOCK;
```

Comment:
Neither of these flags are set when the hande was opened as overlapped, which is the Windows equivalent of nonblocking.

Reply:
Does the opposite implication hold? In other words, does a file which hasn't been opened as overlapped necessarily contain any of these flags?

Anyway, we may want to temporarily put this discussion on hold until we decide if we want to support `O_NONBLOCK` for regular files at all, as per my other comment.

Reply:
Yes, when a file is opened without `FILE_FLAG_OVERLAPPED`, it will always have one of those set (typically `FILE_SYNCHRONOUS_IO_NONALERT`).

Reply:
Please add a comment about it, then :)

Reply:
I've removed reporting on this flag at all from `fd_fdstat_get` since we're no longer opening files with `FILE_FLAG_OVERLAPPED`, so no comment was necessary.